map/bookmark_helpers.cpp

```diff
@@ -370,46 +371,49 @@ std::string GetKMLorGPXPath(std::string const & filePath)
     if (kmlData == nullptr)
       return {};
 
-    fileSavePath = GenerateValidAndUniqueFilePathForKML(base::FileNameFromFullPath(filePath));
+    auto fileSavePath = GenerateValidAndUniqueFilePathForKML(base::FileNameFromFullPath(filePath));
     if (!SaveKmlFileByExt(*kmlData, fileSavePath))
       return {};
+    return {std::move(fileSavePath)};
   }
   else if (fileExt == kKmzExtension)
   {
-    // Extract KML file from KMZ archive and save to temp KML with unique name.
+    // Extract KML files from KMZ archive and save to temp KMLs with unique name.
+    std::vector<std::string> kmlFilePaths;
     try
     {
       ZipFileReader::FileList files;
       ZipFileReader::FilesList(filePath, files);
-      std::string kmlFileName;
-      std::string ext;
-      for (size_t i = 0; i < files.size(); ++i)
+      std::string indexFileName;
+      std::vector<std::string> kmlFileNames;
+      files.erase(std::remove_if(files.begin(), files.end(),
+          [](auto const & file){ return GetLowercaseFileExt(file.first) != kKmlExtension; }),
+          files.end());
+
+      for (const auto & file : files)
       {
-        ext = GetLowercaseFileExt(files[i].first);
-        if (ext == kKmlExtension)
-        {
-          kmlFileName = files[i].first;
-          break;
-        }
+        auto kmlFileInZip = file.first;
+        auto const name = base::FileNameFromFullPath(kmlFileInZip);
+        // Skip index kml if there are more than 1 kml file in kmz.
+        // We assume that there is no useful information there (tracks/bookmarks data is in other kml files).
+        if (name == "doc.kml" && files.size() > 1)
```

Comment:
The robust code should detect the first kml file in the zip archive's root folder, not relying on the specific file name like doc.kml.

Reply:
As 99% of files will have only one kml in the zip, it makes sense to explicitly check if there is only one file and process it accordingly, falling back to more than one file when necessary.

Also it would make sense to find a reliable way of finding and ignoring the first index file in the archive. For example, will the zip listing function return a proper first file if a zip can be created by adding folders first, and then index file? Would it be enough to detect the index file by the absence of slashes in file names inside the archive?

Reply:
I didn't get an idea  about 'first kml file' that we need to skip. I don't see a requirement that referenced kmls should be in some subfolder - so I think we can have doc.kml (index) with links to aaa.kml and zzz.kml in the same folder, so in this case filtering by name is more reliable. 
If you think that we should not use name as a hint maybe I should fix this place:
map/bookmark_manager.cpp:1861 (BookmarkManager::LoadBookmarkRoutine)

```
      base::DeleteFileX(fileToLoad);
      if (kmlData)
      {
        auto kmlFileToLoad = GenerateValidAndUniqueFilePathForKML(base::GetNameFromFullPathWithoutExt(fileToLoad));
```

so we don't just check kmlData for null, but also check that kmlData->m_bookmarksData or kmlData->m_bookmarksData are not empty?

Reply:
> Put the default KML file (doc.kml, or whatever name you want to give it) at the top level within this folder. Include only one .kml file. (When Google Earth opens a KMZ file, it scans the file, looking for the first .kml file in this list. It ignores all subsequent .kml files, if any, in the archive. If the archive contains multiple .kml files, you cannot be sure which one will be found first, so you need to include only one.) 

We can rely on the following logic:
1. Get all kml files in the root dir of the archive.
2. Interpret the first of them, returned by the zip API, as the main file. And interpret any other files as non-main ones.

Reply:
In kmz file used as test all files are not in the root zip folder, but in BACRNKMZ.
I think it could be not reliable to use just first file returned by library (while it was a correct file for my test) and I tried to solve it using fetching file with minimal number of "/" in path - I added it with last commit and test works now.
What do you think about it?

Reply:
Did you create that test file manually or get from some other source?

Reply:
But it has the doc.kml file in the root. 

For kmz without a kml file in the root we can either assume that the first one returned by the zip list function is an index file, or (more robust) we can try to parse/load its contents and see if it is empty or not.

Reply:
Hm, the order can be random. Then the robust logic can try to import the KML and see if it's empty to skip (or we can even support some special index files, for example, for custom icons backup/import).

Reply:
I removed filter from GetKMLOrGPXFilesPathsToLoad and added check for empty to BookmarkManager::LoadBookmarkRoutine. Please have a look.